### PR TITLE
fix: don't create aws_kms keys when in check mode

### PR DESCRIPTION
**Pull request by Tyler-2**
##### SUMMARY

Fixes https://github.com/ansible/ansible/issues/68019

https://github.com/ansible/ansible/issues/68019

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

aws_kms

##### ADDITIONAL INFORMATION

When running aws_kms for key creation in check mode, check mode is ignored and the key is actually created. This commit solves this. See issue description for more details.

```
(venv) tyler@func-platform:~/Repos/ansible-fork$ ansible-playbook tyler-test.yml --check
TASK [Create KMS key] *********************************************************************************************************************************************************************************************
task path: /home/tyler/Repos/ansible-fork/tyler-test.yml:5
changed: [localhost] => {
    "changed": true, 
    "invocation": {
        "module_args": {
            "alias": "test-kms-key", 
            "aws_access_key": null, 
            "aws_config": null, 
            "aws_secret_key": null, 
            "debug_botocore_endpoint_logs": false, 
            "description": null, 
            "ec2_url": null, 
            "enable_key_rotation": null, 
            "enabled": true, 
            "grants": [], 
            "key_id": null, 
            "policy": null, 
            "policy_clean_invalid_entries": true, 
            "policy_grant_types": null, 
            "policy_mode": "grant", 
            "policy_role_arn": null, 
            "policy_role_name": null, 
            "profile": null, 
            "purge_grants": false, 
            "purge_tags": false, 
            "region": null, 
            "security_token": null, 
            "state": "present", 
            "tags": {}, 
            "validate_certs": true
        }
    }
}
META: ran handlers
META: ran handlers

PLAY RECAP ********************************************************************************************************************************************************************************************************
localhost                  : ok=1    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   

(venv) tyler@func-platform:~/Repos/ansible-fork$ aws kms list-aliases | grep test-kms-key
(venv) tyler@func-platform:~/Repos/ansible-fork$ 
```

### CODE REVIEW

**Comment by Tyler-2** [comment]
@tremble I think I have added sane ftests (though I'm not sure where they run, since it doesn't look like shippable knows about them). Also made the change to avoid the big conditional block, since that was trivial.

**Comment by Tyler-2** [comment]
Checking into ftest failure

File: plugins/modules/aws_kms.py [hunk 1]
```diff
@@ -821,18 +821,19 @@ def create_key(connection, module):
     if module.params.get('policy'):
         params['Policy'] = module.params['policy']
 
+    if module.check_mode:
```
Comment (flowerysong) [inline plugins/modules/aws_kms.py:824]: This should go at the very top of the function.

File: plugins/modules/aws_kms.py [hunk 2]
```diff
@@ -821,18 +821,19 @@ def create_key(connection, module):
...
+        return {'changed': True}
+
     try:
         result = connection.create_key(**params)['KeyMetadata']
     except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
         module.fail_json_aws(e, msg="Failed to create initial key")
     key = get_key_details(connection, module, result['KeyId'])
-
```
Comment (flowerysong) [inline plugins/modules/aws_kms.py:829]: I don't think this blank line should be removed.

File: plugins/modules/aws_kms.py [hunk 3]
```diff
@@ -821,18 +821,19 @@ def create_key(connection, module):
...
     update_alias(connection, module, key, module.params['alias'])
     update_key_rotation(connection, module, key, module.params.get('enable_key_rotation'))
 
     ensure_enabled_disabled(connection, module, key, module.params.get('enabled'))
     update_grants(connection, module, key, module.params.get('grants'), False)
-
```
Comment (flowerysong) [inline plugins/modules/aws_kms.py:835]: Same here.

File: tests/integration/targets/aws_kms/tasks/main.yml [hunk 4]
```diff
@@ -32,6 +32,26 @@
...
 
+    - name: create a key in check mode
+      check_mode: yes
+      aws_kms:
+        alias: "{{ resource_prefix }}-kms-check"
+        tags:
+          Hello: World
+        state: present
+        enabled: yes
+
```
Comment (flowerysong) [inline tests/integration/targets/aws_kms/tasks/main.yml:43]: Needs an `assert` task to verify that the module correctly returned 'changed'.

File: tests/integration/targets/aws_kms/tasks/main.yml [hunk 5]
```diff
@@ -65,6 +85,13 @@
           - create_kms.tags['Hello'] == 'World'
           - create_kms.enable_key_rotation == true
 
+    - name: delete the key in check mode
+      check_mode: yes
+      aws_kms:
+        alias: "{{ resource_prefix }}-kms"
+        state: absent
```
Comment (flowerysong) [inline tests/integration/targets/aws_kms/tasks/main.yml:92]: Also needs an `assert`.

File: plugins/modules/aws_kms.py (hunk 1 above)
Comment (Tyler-2) [reply plugins/modules/aws_kms.py:824]: I kind of like to run as much of the module as possible, up until the thing that actually does the change. Is this unconventional with check mode?

File: plugins/modules/aws_kms.py (hunk 2 above)
Comment (Tyler-2) [reply plugins/modules/aws_kms.py:829]: Inadvertent when I made the restruct change. Corrected.

File: plugins/modules/aws_kms.py (hunk 3 above)
Comment (Tyler-2) [reply plugins/modules/aws_kms.py]: Fixed.

File: tests/integration/targets/aws_kms/tasks/main.yml (hunk 5 above)
Comment (Tyler-2) [reply tests/integration/targets/aws_kms/tasks/main.yml:94]: Addressed.

File: tests/integration/targets/aws_kms/tasks/main.yml (hunk 4 above)
Comment (Tyler-2) [reply tests/integration/targets/aws_kms/tasks/main.yml:44]: Addressed.

**Comment by Tyler-2** [comment]
@flowerysong I think I addressed most of your concerns, but I generally let the requestor hit the "resolve conversations". Let me know if there's something more.

File: plugins/modules/aws_kms.py (hunk 1 above)
Comment (tremble) [reply plugins/modules/aws_kms.py:824]: depends on the author, I tend to keep it as 'late' as possible too.

Comment (flowerysong) [reply plugins/modules/aws_kms.py:824]: In this particular case those few lines don't actually do anything, so I prefer having it at the beginning to make it clear that the function doesn't do anything.

**Comment by flowerysong** [comment]
@Tyler-2 People don't normally have to ability to resolve those for you, unless they have extra privileges on the repository itself.

I don't.

File: plugins/modules/aws_kms.py (hunk 1 above)
Comment (Tyler-2) [reply plugins/modules/aws_kms.py]: Made this change.

**Pull request merged**